### PR TITLE
REF: use shareable code for DTI/TDI.insert

### PR DESCRIPTION
**Pull request by jbrockmendel**
xref #30757 should go in before this because it contains the tests.  After this, we'll be able to de-duplicate the two methods.

### CODE REVIEW

**Review by WillAyd — APPROVED** [review]
lgtm. TypeError makes more sense than ValueError where changed I think

**Comment by TomAugspurger** [comment]
Will need an API-breaking whatsnew for changing the type of the exception.

Though to confirm: TypeErrors are typically raised when the argument has an incorrect type. In this case, the type looks correct, since it's a `Timestamp`. It's a tz-naive timestamp though, so I wouldn't expect a ValueError.

File: pandas/tests/indexing/test_coercion.py [hunk 1]
```diff
@@ -432,13 +432,19 @@ def test_insert_index_datetimes(self, fill_val, exp_dtype):
         )
         self._assert_insert_conversion(obj, fill_val, exp, exp_dtype)
 
-        msg = "Passed item and index have different timezone"
         if fill_val.tz:
-            with pytest.raises(ValueError, match=msg):
+            msg = "Cannot compare tz-naive and tz-aware"
```
Comment (jreback) [inline pandas/tests/indexing/test_coercion.py]: @TomAugspurger is this what you are talking about? this I think was just wrong before, not an API change.

Comment (TomAugspurger) [reply pandas/tests/indexing/test_coercion.py]: Yeah. I see now that we raise TypeError for `Timestamp == Timestamp`, which I think is incorrect but not worth changing.

I still thing think this merits a release note.

Comment (jreback) [reply pandas/tests/indexing/test_coercion.py]: sure, @jbrockmendel can you add for what is the user viisble change.

Comment (jbrockmendel) [reply pandas/tests/indexing/test_coercion.py]: updated with whatsnew + green

**Comment by jreback** [comment]
thanks @jbrockmendel

**Pull request merged**